app/src/citations.py

```diff
@@ -220,20 +303,27 @@ def move_citation(match: Match) -> str:
     return re.sub(r" *(( *\(citation-\d+\))+) *([\.\?\!])", move_citation, response).strip()
 
 
-@dataclass
-class ResponseWithSubsections:
+class ResponseWithSubsections(NamedTuple):
     response: str
     subsections: Sequence[Subsection]
 
 
-def simplify_citation_numbers(result: ResponseWithSubsections) -> ResponseWithSubsections:
+def simplify_citation_numbers(
+    response: str, subsections: Sequence[Subsection]
+) -> ResponseWithSubsections:
     """
     Returns the response with remapped `(citation-X)` strings and
     a list of subsections representing the citations.
     The returned subsections only contain citations used in the response
     and are ordered consecutively starting from 1.
     """
-    formatted_response = move_citations_after_punctuation(result.response)
-    remapped_citations = remap_citation_ids(result.subsections, formatted_response)
-    remapped_response = replace_citation_ids(formatted_response, remapped_citations)
+    formatted_response = move_citations_after_punctuation(response)
+
+    merged_subsection_response, merged_subsections = merge_contiguous_cited_subsections(
+        formatted_response, subsections
+    )
```

Comment:
This change affects callers of `simplify_citation_numbers()`, including `chat_api` and `batch_processing`.